mule-extensions-api-persistence/src/main/java/org/mule/extension/api/persistence/ModelPropertyMapTypeAdapter.java

```diff
@@ -0,0 +1,131 @@
+ * ClassLoader the {@link ModelProperty} object will be discarded
+ *
+ * @since 1.0
+ */
+final class ModelPropertyMapTypeAdapter extends TypeAdapter<Map<Class<? extends ModelProperty>, ModelProperty>>
+{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ModelPropertyMapTypeAdapter.class);
+    private static final Map<Class<? extends ModelProperty>, String> classNameMapping = PersistenceConstants.getClassNameMapping();
+    private static final Map<String, Class<? extends ModelProperty>> nameClassMapping = PersistenceConstants.getNameClassMapping();
+
+    private final Gson gson;
+
+    ModelPropertyMapTypeAdapter(Gson gson)
+    {
+        this.gson = gson;
+    }
+
+    @Override
+    public void write(JsonWriter out, Map<Class<? extends ModelProperty>, ModelProperty> modelPropertyMap) throws IOException
+    {
+        out.beginObject();
+        for (Map.Entry<Class<? extends ModelProperty>, ModelProperty> modelProperty : modelPropertyMap.entrySet())
+        {
+            if (modelProperty.getValue().isExternalizable())
+            {
+                out.name(getSerializableModelPropertyName(modelProperty.getKey()));
+                final TypeAdapter adapter = gson.getAdapter(modelProperty.getKey());
+                adapter.write(out, modelProperty.getValue());
+            }
+        }
+        out.endObject();
+
+    }
+
+    @Override
+    public Map<Class<? extends ModelProperty>, ModelProperty> read(JsonReader in) throws IOException
+    {
+        final HashMap<Class<? extends ModelProperty>, ModelProperty> modelPropertyHashMap = new HashMap<>();
+
+        in.beginObject();
+        while (in.hasNext())
+        {
+            final Optional<Class<? extends ModelProperty>> modelPropertyClass = getClassForModelProperty(in.nextName());
+
+            if (modelPropertyClass.isPresent())
+            {
+                final Class<? extends ModelProperty> type = modelPropertyClass.get();
+                final TypeAdapter<?> adapter = gson.getAdapter(type);
+                final ModelProperty read = (ModelProperty) adapter.read(in);
+                modelPropertyHashMap.put(type, read);
+            }
+        }
+        in.endObject();
+        return modelPropertyHashMap;
+    }
+
+    private Optional<Class<? extends ModelProperty>> getClassForModelProperty(String modelPropertyName)
+    {
+        Class<? extends ModelProperty> modelPropertyClass = null;
+        if (nameClassMapping.containsKey(modelPropertyName))
+        {
+            modelPropertyClass = nameClassMapping.get(modelPropertyName);
+        }
+        else
+        {
+            try
+            {
+                modelPropertyClass = (Class<? extends ModelProperty>) Class.forName(modelPropertyName);
+            }
+            catch (ClassNotFoundException e)
+            {
+                LOGGER.error("Error loading ModelProperty. Class not found in the current classloader", e);
```

Comment:
deserealization should fail rather than just logging an error. Throw an exception here